compiler/noirc_frontend/src/parser/parser/generics.rs

```diff
@@ -0,0 +1,295 @@
+use noirc_errors::Span;
+
+use crate::{
+    ast::{
+        GenericTypeArg, GenericTypeArgs, IntegerBitSize, Signedness, UnresolvedGeneric,
+        UnresolvedGenerics, UnresolvedType, UnresolvedTypeData,
+    },
+    parser::{labels::ParsingRuleLabel, ParserErrorReason},
+    token::{Keyword, Token, TokenKind},
+};
+
+use super::{parse_many::separated_by_comma, Parser};
+
+impl<'a> Parser<'a> {
+    /// Generics = ( '<' GenericsList? '>' )?
+    ///
+    /// GenericsList = Generic ( ',' Generic )* ','?
+    pub(super) fn parse_generics(&mut self) -> UnresolvedGenerics {
+        if !self.eat_less() {
+            return Vec::new();
+        }
+
+        self.parse_many(
+            "generic parameters",
+            separated_by_comma().until(Token::Greater),
+            Self::parse_generic_in_list,
+        )
+    }
+
+    fn parse_generic_in_list(&mut self) -> Option<UnresolvedGeneric> {
+        if let Some(generic) = self.parse_generic() {
+            Some(generic)
+        } else {
+            self.expected_label(ParsingRuleLabel::GenericParameter);
+            None
+        }
+    }
+
+    /// Generic
+    ///     = VariableGeneric
+    ///     | NumericGeneric
+    ///     | ResolvedGeneric
+    fn parse_generic(&mut self) -> Option<UnresolvedGeneric> {
+        if let Some(generic) = self.parse_variable_generic() {
+            return Some(generic);
+        }
+
+        if let Some(generic) = self.parse_numeric_generic() {
+            return Some(generic);
+        }
+
+        if let Some(generic) = self.parse_resolved_generic() {
+            return Some(generic);
+        }
+
+        None
+    }
+
+    /// VariableGeneric = identifier
+    fn parse_variable_generic(&mut self) -> Option<UnresolvedGeneric> {
+        if let Some(ident) = self.eat_ident() {
+            return Some(UnresolvedGeneric::Variable(ident));
+        }
+
+        None
+    }
+
+    /// NumericGeneric = 'let' identifier ':' Type
+    fn parse_numeric_generic(&mut self) -> Option<UnresolvedGeneric> {
+        if !self.eat_keyword(Keyword::Let) {
+            return None;
+        }
+
+        let Some(ident) = self.eat_ident() else {
+            return None;
+        };
+
+        if !self.eat_colon() {
+            self.push_error(
+                ParserErrorReason::MissingTypeForNumericGeneric,
+                self.current_token_span,
+            );
+            return Some(UnresolvedGeneric::Numeric { ident, typ: type_u32() });
+        }
+
+        let typ = self.parse_type_or_error();
+        if let UnresolvedTypeData::Integer(signedness, bit_size) = &typ.typ {
+            if matches!(signedness, Signedness::Signed)
+                || matches!(bit_size, IntegerBitSize::SixtyFour)
+            {
+                self.push_error(ParserErrorReason::ForbiddenNumericGenericType, typ.span);
+            }
+        }
+
+        Some(UnresolvedGeneric::Numeric { ident, typ })
```

Comment:
We only allow integer types here (not sure if we also allow field or not) so we should probably error if this is not an integer? Although we already do this during name resolution IIRC.

Reply:
Right, I thought so too. I initially errored on other types but some test failed so I removed that. It's also the same in the chumsky parser: https://github.com/noir-lang/noir/blob/8236cbdff60c1aaf41fc53142b6f0f9ea2fc2fa8/compiler/noirc_frontend/src/parser/parser/function.rs#L129-L143